interfaces/prompting/constraints.go

```diff
@@ -30,62 +31,93 @@ import (
+func (c *Constraints) Match(path string) (bool, error) {
 	if c.PathPattern == nil {
-		return prompting_errors.NewInvalidPathPatternError("", "no path pattern")
+		return false, prompting_errors.NewInvalidPathPatternError("", "no path pattern")
+	}
+	match, err := c.PathPattern.Match(path)
+	if err != nil {
+		// Error should not occur, since it was parsed internally
+		return false, prompting_errors.NewInvalidPathPatternError(c.PathPattern.String(), err.Error())
 	}
-	return c.validatePermissions(iface)
+	return match, nil
 }
 
-// validatePermissions checks that the permissions for the given constraints
-// are valid for the given interface. If not, returns an error, otherwise
-// ensures that the permissions are in the order in which they occur in the
-// list of available permissions for that interface.
-func (c *Constraints) validatePermissions(iface string) error {
-	availablePerms, ok := interfacePermissionsAvailable[iface]
-	if !ok {
-		return prompting_errors.NewInvalidInterfaceError(iface, availableInterfaces())
-	}
-	permsSet := make(map[string]bool, len(c.Permissions))
-	var invalidPerms []string
-	for _, perm := range c.Permissions {
-		if !strutil.ListContains(availablePerms, perm) {
-			invalidPerms = append(invalidPerms, perm)
-			continue
+// ContainPermissions returns true if the permission map in the constraints
+// includes every one of the given permissions.
+//
+// This method is only intended to be called on constraints which have just
+// been created from a reply, to check that the reply covers the request.
+func (c *Constraints) ContainPermissions(permissions []string) bool {
+	for _, perm := range permissions {
+		if _, exists := c.Permissions[perm]; !exists {
+			return false
 		}
-		permsSet[perm] = true
 	}
-	if len(invalidPerms) > 0 {
-		return prompting_errors.NewInvalidPermissionsError(iface, invalidPerms, availablePerms)
+	return true
+}
+
+// ToRuleConstraints validates the receiving Constraints and converts it to
+// RuleConstraints. If the constraints are not valid with respect to the given
+// interface, returns an error.
+func (c *Constraints) ToRuleConstraints(iface string, currTime time.Time) (*RuleConstraints, error) {
+	if c.PathPattern == nil {
+		return nil, prompting_errors.NewInvalidPathPatternError("", "no path pattern")
 	}
-	if len(permsSet) == 0 {
-		return prompting_errors.NewPermissionsListEmptyError(iface, availablePerms)
+	rulePermissions, err := c.Permissions.toRulePermissionMap(iface, currTime)
+	if err != nil {
+		return nil, err
 	}
-	newPermissions := make([]string, 0, len(permsSet))
-	for _, perm := range availablePerms {
-		if exists := permsSet[perm]; exists {
-			newPermissions = append(newPermissions, perm)
-		}
+	ruleConstraints := &RuleConstraints{
+		PathPattern: c.PathPattern,
+		Permissions: rulePermissions,
 	}
-	c.Permissions = newPermissions
-	return nil
+	return ruleConstraints, nil
+}
+
+// RuleConstraints hold information about the applicability of an existing rule
```

Comment:
feels like this should be name ConstraintsRule?

Reply:
no, that doesn't sound right either.  Constraints that go into a rule seems fine to be called RuleConstraints, I understand that it reads also as Constraints for a Rule which might mean a couple of things, but I think contextually is clear enough